_layouts/posts.html

```diff
@@ -27,15 +27,29 @@
                     <a href="{{ post.url }}"><p>{{ post.title | escape }}</p></a>
                 {% endif %}
                 </div>
-
-                <p class="subtitle">
+                
+                    <p class="subtitle">
                     <time datetime="{{ post.date | date: date_format }}">{{ post.date | date: date_format }}</time>
                     - {{ authors | remove_first: ', ' | strip }}
-                </p>
+                    </p> 
+                
+        <div class="post-categories">
+            {% if post %}
+              {% assign categories = post.categories %}
+            {% else %}
+              {% assign categories = page.categories %}
+            {% endif %}
```

Comment:
I would just focus on the `post.categories`, remove the conditional test there, and use `page.categories` directly in the for loop